source/neuropod/multiprocess/mq/ipc_message_queue_impl.hh

```diff
@@ -152,7 +162,32 @@ void IPCMessageQueue<UserPayloadType>::send_message(const WireFormat &msg)
         SPDLOG_TRACE("OPE: Sending IPC control message {}.", msg.type);
     }
 
-    send_queue_->send(&msg, sizeof(msg), 0);
+    // Send w/ timeout + heartbeat check
+    while (true)
+    {
+        auto timeout_at = boost::interprocess::microsec_clock::universal_time() +
+                          boost::posix_time::milliseconds(detail::HEARTBEAT_INTERVAL_MS);
+
+        if (send_queue_->timed_send(&msg, sizeof(msg), 0, timeout_at))
+        {
+            // Successfully sent
+            break;
+        }
+
+        // Make sure the worker process is still alive
+        throw_if_lost_heartbeat();
+    }
+}
+
+// Throw an error if we lost communication with the other process
+template <typename UserPayloadType>
+void IPCMessageQueue<UserPayloadType>::throw_if_lost_heartbeat()
+{
+    if (lost_heartbeat_.load(std::memory_order_relaxed))
+    {
+        // We lost the heartbeat so let's throw an error here (where the user can catch it)
+        NEUROPOD_ERROR("OPE lost communication with the other process. See logs for more details.");
```

Comment:
This can be considered separately, just mention here. This will throw exception:

```
 template <typename... Params>
 [[noreturn]] void throw_error(const char *file, int line, const char *function, Params &&... params)
 {
     // Log the error
     // (this is effectively what the SPDLOG_ERROR macro expands to)
     spdlog::default_logger_raw()->log(
         spdlog::source_loc{file, line, function}, spdlog::level::err, std::forward<Params>(params)...);

     // Throw a runtime error
     throw std::runtime_error("Neuropod Error: " + fmt::format(std::forward<Params>(params)...));
 }
 ```
 
 If parent/core want to distinguish exceptions, this will need to parse/compare exception message. It would be useful to have some way to categorize it, I don't think that attribute "category" is necessary but maybe having some "tags" as a contract may help.

Reply:
Makes sense - I can make the exception distinguishable